pkg/executor/join/hash_join_stats.go

```diff
@@ -0,0 +1,261 @@
+	if !ok {
+		return
+	}
+	e.fetchAndBuildHashTable += tmp.fetchAndBuildHashTable
+	e.hashStat.buildTableElapse += tmp.hashStat.buildTableElapse
+	e.hashStat.probeCollision += tmp.hashStat.probeCollision
+	e.fetchAndProbe += tmp.fetchAndProbe
+	e.probe += tmp.probe
+	if e.maxFetchAndProbe < tmp.maxFetchAndProbe {
+		e.maxFetchAndProbe = tmp.maxFetchAndProbe
+	}
+}
+
+type hashStatistic struct {
+	// NOTE: probeCollision may be accessed from multiple goroutines concurrently.
+	probeCollision   int64
+	buildTableElapse time.Duration
+}
+
+func (s *hashStatistic) String() string {
+	return fmt.Sprintf("probe_collision:%v, build:%v", s.probeCollision, execdetails.FormatDuration(s.buildTableElapse))
+}
+
+type hashJoinRuntimeStatsV2 struct {
+	concurrent     int
+	probeCollision int64
+
+	fetchAndBuildHashTable int64
+
+	partitionData  int64
+	buildHashTable int64
+	probe          int64
+	fetchAndProbe  int64
+
+	maxPartitionData  int64
+	maxBuildHashTable int64
+	maxProbe          int64
+	maxFetchAndProbe  int64
+
+	maxPartitionDataForCurrentRound  int64
+	maxBuildHashTableForCurrentRound int64
+	maxProbeForCurrentRound          int64
+	maxFetchAndProbeForCurrentRound  int64
+}
+
+func setMaxValue(addr *int64, currentValue int64) {
+	for {
+		value := atomic.LoadInt64(addr)
+		if currentValue <= value {
+			return
+		}
+		if atomic.CompareAndSwapInt64(addr, value, currentValue) {
+			return
+		}
+	}
+}
+
+func (e *hashJoinRuntimeStatsV2) reset() {
+	e.probeCollision = 0
+	e.fetchAndBuildHashTable = 0
+	e.partitionData = 0
+	e.buildHashTable = 0
+	e.probe = 0
+	e.fetchAndProbe = 0
+	e.maxPartitionData = 0
+	e.maxBuildHashTable = 0
+	e.maxProbe = 0
+	e.maxFetchAndProbe = 0
+	e.maxPartitionDataForCurrentRound = 0
+	e.maxBuildHashTableForCurrentRound = 0
+	e.maxProbeForCurrentRound = 0
+	e.maxFetchAndProbeForCurrentRound = 0
+}
+
+func (e *hashJoinRuntimeStatsV2) resetCurrentRound() {
+	e.maxPartitionData += e.maxPartitionDataForCurrentRound
+	e.maxBuildHashTable += e.maxBuildHashTableForCurrentRound
+	e.maxProbe += e.maxProbeForCurrentRound
+	e.maxFetchAndProbe += e.maxFetchAndProbeForCurrentRound
+	e.maxPartitionDataForCurrentRound = 0
+	e.maxBuildHashTableForCurrentRound = 0
+	e.maxProbeForCurrentRound = 0
+	e.maxFetchAndProbeForCurrentRound = 0
+}
+
+// Tp implements the RuntimeStats interface.
+func (*hashJoinRuntimeStatsV2) Tp() int {
+	return execdetails.TpHashJoinRuntimeStatsV2
+}
+
+func (e *hashJoinRuntimeStatsV2) String() string {
+	buf := bytes.NewBuffer(make([]byte, 0, 128))
+	if e.fetchAndBuildHashTable > 0 {
+		buf.WriteString("build_hash_table:{total:")
+		buf.WriteString(execdetails.FormatDuration(time.Duration(e.fetchAndBuildHashTable)))
+		buf.WriteString(", fetch:")
+		buf.WriteString(execdetails.FormatDuration(time.Duration(int64(e.fetchAndBuildHashTable) - e.maxBuildHashTable - e.maxPartitionData)))
+		buf.WriteString(", build:")
+		buf.WriteString(execdetails.FormatDuration(time.Duration(e.maxBuildHashTable)))
```

Comment:
I think for current version, you can use `maxBuildHashTable + maxPartitionData` as the build time.

Reply:
> I think for current version, you can use `maxBuildHashTable + maxPartitionData` as the build time.

done